core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java

```diff
@@ -134,36 +137,28 @@ public void runPhase(final ModuleFlowProcessingPhaseTemplate template, final Mes
       // TODO MULE-11167 Policies should be non blocking
       if (!enableSourcePolicies) {
         just(templateEvent)
-            .doOnNext(request -> {
-              fireNotification(messageProcessContext.getMessageSource(), request,
-                               messageProcessContext.getFlowConstruct(),
-                               MESSAGE_RECEIVED);
-            })
-            .then(request -> from(template.routeEventAsync(request)))
-            .doOnSuccess(getSuccessConsumer(messageSource, templateEvent, messageProcessContext, phaseResultNotifier, template,
-                                            terminateConsumer))
-            .doOnError(MessagingException.class, me -> {
-              if (me.getCause() instanceof SourceErrorException
-                  && sourceResponseErrorTypeMatcher.match(((SourceErrorException) me.getCause()).getErrorType())) {
-                try {
-                  handleSourceError(exceptionHandler, errorConsumer, (SourceErrorException) me.getCause());
-                  onTerminate(terminateConsumer, left(me.getEvent()));
-                } catch (Exception e) {
-                  onTerminate(terminateConsumer, right(e));
-                }
+            .doOnNext(request -> fireNotification(messageProcessContext.getMessageSource(), request,
+                                                  messageProcessContext.getFlowConstruct(),
+                                                  MESSAGE_RECEIVED))
+            .then(request -> from(template.routeEventAsync(request)).switchIfEmpty(fromCallable(() -> emptyEvent(templateEvent))))
+            .then(onSuccess(messageSource, templateEvent, messageProcessContext, phaseResultNotifier, template,
+                            terminateConsumer))
+            .onErrorResume(SourceErrorException.class,
+                           see -> onSourceException(exceptionHandler, errorHandler, terminateConsumer, see))
+            .onErrorResume(MessagingException.class, me -> {
+              if (me.getCause() instanceof SourceErrorException) {
+                return onSourceException(exceptionHandler, errorHandler, terminateConsumer, (SourceErrorException) me.getCause());
               } else {
-                try {
-                  errorConsumer.accept(me);
-                  onTerminate(terminateConsumer, right(me));
-                } catch (Exception e) {
-                  onTerminate(terminateConsumer, right(e));
-                }
+                return from(errorHandler.apply(me))
+                    .doOnSuccess(v -> onTerminate(terminateConsumer, right(me)))
```

Comment:
onTerminate should be being invoked from main Mono, not in onSucess/onError and here independently.  This should simply be:
`. onErrorResume(me -> errorHandler.apply(me))`